src/Orleans/Runtime/RequestContext.cs

```diff
@@ -112,19 +113,23 @@ public static void Import(Dictionary<string, object> contextData)
                 if (contextData == null || !contextData.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj))
                 {
                     activityIdObj = Guid.Empty;
-                }
 
-#if NETSTANDARD
-                ActivityId.Value = (Guid) activityIdObj;
-#else
-                Trace.CorrelationManager.ActivityId = (Guid)activityIdObj;
+                    object legacyActivityIdObj;
+#if !NETSTANDARD
+                    if (contextData.TryGetValue(E2_E_TRACING_LEGACY_ACTIVITY_ID_HEADER, out legacyActivityIdObj))
+                    {
+                        Trace.CorrelationManager.ActivityId = (Guid)legacyActivityIdObj;
+                    }
 #endif
+                }
+
+                ActivityId.Value = (Guid)activityIdObj;
```

Comment:
It's not expensive from memory standpoint, but ```ExecutionContext``` is optimized for no values case, so flowing of that value will indeed take some CPU time, but I doubt that it could be seen on performance tests. Anyway, fixed.